src/subscription.rs

```diff
@@ -112,70 +181,147 @@ impl fmt::Display for Cancelled {
+pub struct SubscriptionFuture<TRes: Debug + PartialEq> {
+    /// The unique identifier of the subscription request, matching the one in
+    /// the `SubscriptionRegistry`.
+    id: u64,
+    /// The collection of all the live subscriptions.
+    subscriptions: Arc<Mutex<Subscriptions<TRes>>>,
 }
 
-impl<TResult: Clone> Future for SubscriptionFuture<TResult> {
-    type Output = Result<TResult, Cancelled>;
+impl<TRes: Debug + PartialEq> Future for SubscriptionFuture<TRes> {
+    type Output = Result<TRes, Cancelled>;
 
     fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
-        let mut subscription = self.subscription.lock().unwrap();
+        let mut subscription = {
+            // don't hold the lock for too long, otherwise the `Drop` impl for `SubscriptionFuture`
+            // can cause a stack overflow
+            let mut subscriptions = task::block_on(async { self.subscriptions.lock().await });
+            if let Some(sub) = subscriptions.remove(&self.id) {
+                sub
+            } else {
+                // the subscription must already have been cancelled
+                return Poll::Ready(Err(Cancelled));
+            }
+        };
 
-        match &mut *subscription {
+        match subscription {
             Subscription::Cancelled => Poll::Ready(Err(Cancelled)),
-            Subscription::Pending { ref mut wakers } => {
-                let waker = context.waker();
-                if !wakers.iter().any(|w| w.will_wake(waker)) {
-                    wakers.push(waker.clone());
-                }
+            Subscription::Pending { ref mut waker, .. } => {
+                *waker = Some(context.waker().clone());
+                task::block_on(async { self.subscriptions.lock().await })
+                    .insert(self.id, subscription);
                 Poll::Pending
             }
-            Subscription::Ready(result) => Poll::Ready(Ok(result.clone())),
+            Subscription::Ready(result) => Poll::Ready(Ok(result)),
         }
     }
 }
 
-impl<TResult> fmt::Debug for SubscriptionFuture<TResult> {
+impl<TRes: Debug + PartialEq> Drop for SubscriptionFuture<TRes> {
+    fn drop(&mut self) {
+        let (sub, is_last) = task::block_on(async {
+            let mut subscriptions = self.subscriptions.lock().await;
+            let sub = subscriptions.remove(&self.id);
+            // check if this is the last subscription to this resource
+            let is_last = !subscriptions.values().any(|s| Some(s) == sub.as_ref());
+
+            (sub, is_last)
+        });
+
+        if let Some(sub) = sub {
+            if let Subscription::Pending { ref request, .. } = sub {
+                debug!("Dropping subscription {}", request.id);
+            }
+            // don't bother updating anything that isn't `Pending`
+            if let mut sub @ Subscription::Pending { .. } = sub {
+                debug!("It was the last subscription to a resource, sending a cancel notification");
```

Comment:
```suggestion
                debug!("It was the last subscription to a resource, sending a cancel notification: {:?}", sub);
```

maybe?

Reply:
wait couldn't these two matches be combined, maybe combined into the upper, maybe into a single match explaining why the other need not to be cancelled, or something like `Some(_) => { /* nothing to cancel for Cancelled or Ready */ }`?

Reply:
the subscription is already described in the debug right above it, and that one properly handles the ugly display of `Cid` (not a straight `Debug`, but a full `Display` impl)

Reply:
ah no, I actually already changed it; in that case I'll do it so it does :laughing: